### PR TITLE
`azurerm_proximity_placement_group` - support `allowed_vm_sizes` and `zone`

### LINKED ISSUE #19379: Add support for Intent and Zone in azurerm_proximity_placement_group
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

In an October 10 2022 Blog post, Microsoft announced two new parameters of Proximity Placement Groups that would make using them more predictable, or at least would alert operators to issues earlier on in the terraform workflow:
https://azure.microsoft.com/en-us/updates/ppgintent/

Please add Intent and Zone parameters to the azurerm_proximity_placement_group resource.

### New or Affected Resource(s)/Data Source(s)

azurerm_proximity_placement_group

### Potential Terraform Configuration

```hcl
resource "azurerm_resource_group" "example" {
  name     = "example-resources"
  location = "West Europe"
}

resource "azurerm_proximity_placement_group" "example" {
  name                = "exampleProximityPlacementGroup"
  location            = azurerm_resource_group.example.location
  resource_group_name = azurerm_resource_group.example.name
  intent              = "Standard_E64s_v4", "Standard_M416ms_v2"
  zone                = "1"

  tags = {
    environment = "Production"
  }
}
```

### References

https://azure.microsoft.com/en-us/updates/ppgintent/

https://github.com/MicrosoftDocs/azure-docs/blob/main/articles/virtual-machines/windows/proximity-placement-groups.md

https://learn.microsoft.com/en-us/cli/azure/ppg?view=azure-cli-latest#az-ppg-create

https://learn.microsoft.com/en-us/dotnet/api/microsoft.azure.management.compute.models.proximityplacementgroup.intent?view=azure-dotnet

https://learn.microsoft.com/en-us/dotnet/api/microsoft.azure.management.compute.models.proximityplacementgroup.zones?view=azure-dotnet

### PR DESCRIPTION
**Pull request by myc2h6o**
Close #19379

### CODE REVIEW

File: internal/services/compute/proximity_placement_group_resource.go [hunk 1]
```diff
@@ -74,8 +95,29 @@ func resourceProximityPlacementGroupCreateUpdate(d *pluginsdk.ResourceData, meta
...
+		// Need to explicitly set an empty slice when updating to empty vm sizes
+		if payload.Properties.Intent == nil {
+			payload.Properties.Intent = &proximityplacementgroups.ProximityPlacementGroupPropertiesIntent{}
+		}
+		vmSizes := make([]string, 0)
+		payload.Properties.Intent.VMSizes = &vmSizes
+	}
+
+	if v, ok := d.GetOk("zone"); ok {
+		payload.Zones = &[]string{
```
Comment (myc2h6o) [inline internal/services/compute/proximity_placement_group_resource.go:118]: only one zone could be specified though the property name is zones

File: internal/services/compute/proximity_placement_group_resource.go [hunk 2]
```diff
@@ -47,6 +50,24 @@ func resourceProximityPlacementGroup() *pluginsdk.Resource {
...
+				},
+			},
+
+			"zone": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				RequiredWith: []string{"intent_vm_sizes"},
+				ForceNew:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
```
Comment (tombuildsstuff) [inline internal/services/compute/proximity_placement_group_resource.go:69]: we should be using `commonschema.ZoneSingle` here

File: internal/services/compute/proximity_placement_group_resource.go [hunk 3]
```diff
@@ -47,6 +50,24 @@ func resourceProximityPlacementGroup() *pluginsdk.Resource {
 
 			"location": commonschema.Location(),
 
+			"intent_vm_sizes": {
+				Type:     pluginsdk.TypeSet,
+				Optional: true,
+				MinItems: 1,
+				Elem: &pluginsdk.Schema{
+					Type:         pluginsdk.TypeString,
+					ValidateFunc: validation.StringIsNotWhiteSpace,
```
Comment (tombuildsstuff) [inline internal/services/compute/proximity_placement_group_resource.go:59]: the Compute SDK exposes a list of VM SKU's?

File: internal/services/compute/proximity_placement_group_resource.go [hunk 4]
```diff
@@ -111,6 +153,22 @@ func resourceProximityPlacementGroupRead(d *pluginsdk.ResourceData, meta interfa
...
+				}
+			}
+		}
+
+		zone := ""
+		if model.Zones != nil && len(*model.Zones) > 0 {
+			z := *model.Zones
+			zone = z[0]
+		}
+		d.Set("zone", zone)
```
Comment (tombuildsstuff) [inline internal/services/compute/proximity_placement_group_resource.go:170]: we have common schema logic for zones: https://github.com/hashicorp/go-azure-helpers/tree/main/resourcemanager/zones

File: internal/services/compute/proximity_placement_group_resource.go [hunk 5]
```diff
@@ -111,6 +153,22 @@ func resourceProximityPlacementGroupRead(d *pluginsdk.ResourceData, meta interfa
...
 
 		d.Set("location", location.Normalize(model.Location))
+
+		if props := model.Properties; props != nil {
+			if intent := props.Intent; intent != nil {
+				if intent.VMSizes != nil {
+					d.Set("intent_vm_sizes", set.FromStringSlice(*intent.VMSizes))
+				}
+			}
+		}
```
Comment (tombuildsstuff) [inline internal/services/compute/proximity_placement_group_resource.go:163]: this field always needs to be set into the state

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @myc2h6o 

Thanks for this PR - I've taken a look through and left some comments inline, if we can fix those up then we should be able to take another look.

Thanks!

File: internal/services/compute/proximity_placement_group_resource.go (hunk 2 above)
Comment (myc2h6o) [reply internal/services/compute/proximity_placement_group_resource.go:70]: `zone` in this resource is set to `RequiredWith: []string{"intent_vm_sizes"}`, thus using a separate schema.

File: internal/services/compute/proximity_placement_group_resource.go (hunk 3 above)
Comment (myc2h6o) [reply internal/services/compute/proximity_placement_group_resource.go]: have updated to `virtualmachines.PossibleValuesForVirtualMachineSizeTypes()`

File: internal/services/compute/proximity_placement_group_resource.go (hunk 4 above)
Comment (myc2h6o) [reply internal/services/compute/proximity_placement_group_resource.go]: updated

File: internal/services/compute/proximity_placement_group_resource.go (hunk 5 above)
Comment (myc2h6o) [reply internal/services/compute/proximity_placement_group_resource.go]: updated

**Comment by myc2h6o** [comment]
Hi @tombuildsstuff thanks for reviewing the pr! I've resolved the comments, please take a look.

**Comment by myc2h6o** [comment]
Test result
image

File: website/docs/r/proximity_placement_group.html.markdown [hunk 6]
```diff
@@ -40,6 +40,12 @@ The following arguments are supported:
 
 * `location` - (Required) Specifies the supported Azure location where the resource exists. Changing this forces a new resource to be created.
 
+* `intent_vm_sizes` - (Optional) Specifies the supported sizes of Virtual Machines that can be created in the Proximity Placement Group.
```
Comment (katbyte) [inline website/docs/r/proximity_placement_group.html.markdown:43]: I'm not sure intent is very clear here? would this make more sense as
```suggestion
* `allowed_vm_sizes` - (Optional) Specifies the supported sizes of Virtual Machines that can be created in the Proximity Placement Group.
```

**Review by katbyte — CHANGES_REQUESTED** [review]

File: website/docs/r/proximity_placement_group.html.markdown (hunk 6 above)
Comment (myc2h6o) [reply website/docs/r/proximity_placement_group.html.markdown]: updated

**Review by katbyte — APPROVED** [review]
LGTM 🌵

**Pull request merged**